### PR TITLE
JDK-8266222: [aix] In mmap-mode, partial releases with os::release_memory may trash internal bookkeeping

### PR DESCRIPTION
**Pull request by tstuefe**
In os_aix.cpp we keep book about for reserved address ranges since we need to keep information about which API (mmap or shmat) had been used for this range. (code hint: see vmembk_(add|remove|find)).

When releasing memory via os::release_memory, we remove those ranges from the internal bookkeeping.

However, for mmap() the release may be partial. In that case the internal bookkeeping is not updated correctly: we just remove the record for the whole original address range. 

After that operation, the remainder address range is still mapped from the point of the VM, but the bookkeeping is lost and subsequent calls to os::release_memory or os::commit/uncommit_memory for this range will run into a guarantee.

Partial releases can happen e.g. when, in the process of running os::reserve_memory_aligned(), the extra-aligned pages are unmapped. This is of more concern now with the new Metaspace, since we now reserve with larger alignments (4G+).

This problem was hidden by the fact that we typically run in 64K paged mode where we use SysV shared memory; but if AME is enabled on AIX, 64K pages are disabled, and the VM falls back to 4K pages and using mmap, and runs into this bug.

Edit: Tested on AIX machines with and without AME activated, solves the problems. All SAP nightlies ran, no problems caused by this patch.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266222](https://bugs.openjdk.java.net/browse/JDK-8266222): [aix] In mmap-mode, partial releases with os::release_memory may trash internal bookkeeping

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3765/head:pull/3765` \
`$ git checkout pull/3765`

Update a local copy of the PR: \
`$ git checkout pull/3765` \
`$ git pull https://git.openjdk.java.net/jdk pull/3765/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3765`

View PR using the GUI difftool: \
`$ git pr show -t 3765`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3765.diff">https://git.openjdk.java.net/jdk/pull/3765.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/aix/os_aix.cpp [hunk 1]
```diff
@@ -1989,6 +1989,7 @@ bool os::pd_release_memory(char* addr, size_t size) {
   // Dynamically do different things for mmap/shmat.
   vmembk_t* const vmi = vmembk_find(addr);
   guarantee0(vmi);
+  vmi->assert_is_valid_subrange(addr, size);
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:1992]: Reviewer note: this is unrelated to the problem, I just lifted the assert call from both branches and made it independent from debug/release.

**Review by TheRealMDoerr — APPROVED** [review]
This looks good.

**Comment by tstuefe** [comment]
Danke, Martin!

**Comment by tstuefe** [comment]
Note that since we don't have that many AIX reviewers, and this only concerns AIX code, I go ahead and push. Its our port anyway, if something breaks we will have to fix it.

/integrate
